Make SDK API more Swift-friendly

DmitriyKirakosyan opened the pull request:
<!-- 
Thank you for submitting a pull request! Please add some info (if applicable) to give us some context on the PR.

We will review the PR as soon as possible, leave feedback, add a tag, etc. and let you know what's going on.

Cheers!

The App Center team -->

Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* ~[ ] Did you add unit tests?~
* ~[ ] Did you check UI tests on the sample app? They are not executed on CI.~
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Updates SDK Swift API to make it more _swifty_.

NOTE: I's a breaking change!

## Related PRs or issues

[AB#83267](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/83267)

MatkovIvan commented on AppCenter/AppCenter/Internals/Context/Device/MSACDeviceTracker.h:
```diff
@@ -27,4 +28,18 @@
+ *
+ * @return country code.
+ */
+- (NSString *)getCountryCode;
```
do not use "get" prefix

iamclement commented on AppCenter/AppCenter/MSACAppCenter.h:
```diff
@@ -68,76 +69,45 @@
+ * The flag indicates whether the SDK has already been configured or not.
  */
-+ (BOOL)isConfigured;
+@property(class, atomic, readonly) BOOL isConfigured;
```
ObjC properties should not be prefixed by "is". In ObjC only the getter should be prefixed by "is". Example `@property (assign, getter=isEditable) BOOL editable;` cf [Apple doc](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html).

iamclement commented on AppCenter/AppCenter/MSACAppCenter.h:
```diff
@@ -68,76 +69,45 @@
+ * The flag indicates whether app is running in App Center Test Cloud.
  */
-+ (BOOL)isRunningInAppCenterTestCloud;
+@property(class, atomic, readonly) BOOL isRunningInAppCenterTestCloud;
```
ObjC properties should not be prefixed by "is". In ObjC only the getter should be prefixed by "is". Example `@property (assign, getter=isEditable) BOOL editable;` cf [Apple doc](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html). The good new is it won't break ObjC customers :).

iamclement commented on AppCenter/AppCenter/MSACAppCenter.h:
```diff
@@ -68,76 +69,45 @@
- * @see setEnabled:
  */
-+ (BOOL)isEnabled;
+@property(class, nonatomic, getter=isEnabled, setter=setEnabled:) BOOL enabled NS_SWIFT_NAME(enabled);
```
👍 Right, we don't need the `is` prefix in the Swift property because it's mutable.

iamclement commented on AppCenter/AppCenter/MSACAppCenter.h:
```diff
@@ -151,38 +121,33 @@
  * to `0`. Then you will have to forward any application delegate needed by the SDK manually.
  */
-+ (BOOL)isAppDelegateForwarderEnabled;
+@property(class, readonly, nonatomic) BOOL isAppDelegateForwarderEnabled;
```
ObjC properties should not be prefixed by "is". In ObjC only the getter should be prefixed by "is". Example @property (assign, getter=isEditable) BOOL editable; cf Apple doc. The good new is it won't break ObjC customers :).

iamclement commented on AppCenter/AppCenter/MSACAppCenter.h:
```diff
@@ -151,38 +121,33 @@
- * @return BOOL if the debugger is attached.
  */
-+ (BOOL)isDebuggerAttached;
+@property(class, readonly, nonatomic) BOOL isDebuggerAttached;
```
ObjC properties should not be prefixed by "is". In ObjC only the getter should be prefixed by "is". Example @property (assign, getter=isEditable) BOOL editable; cf Apple doc. The good new is it won't break ObjC customers :).

iamclement reviewed (changes requested):
Naming is not easy thanks for looking at this :). I'll take another look couldn't get to the end.

DmitriyKirakosyan commented on AppCenter/AppCenter/MSACEnable.h:
```diff
@@ -19,7 +20,7 @@
  * @param deleteData A boolean value set to YES to delete data or NO to keep it.
  */
-- (void)setEnabled:(BOOL)isEnabled andDeleteDataOnDisabled:(BOOL)deleteData;
+- (void)setEnabled:(BOOL)isEnabled andDeleteDataOnDisabled:(BOOL)deleteData NS_SWIFT_NAME(setEnabled(_:deleteDataOnDisabled:));
```
Hm.. In this case, do you think it's worth adding `and` for swift too, ie `NS_SWIFT_NAME(setEnabled(_:andDeleteDataOnDisabled:))`?

iamclement replied:
For consistency I would say yes but this API is for internal use so it doesn't really matter.

MatkovIvan commented on AppCenter/AppCenter/MSACAppCenter.h:
```diff
@@ -74,12 +74,12 @@ NS_SWIFT_NAME(AppCenter)
  * The flag indicates whether the SDK has already been configured or not.
  */
-@property(class, atomic, readonly) BOOL isConfigured;
+@property(class, atomic, readonly, getter=isConfigured) BOOL configured NS_SWIFT_NAME(isConfigured);
```
No need for NS_SWIFT_NAME in this case

iamclement commented on AppCenter/AppCenter/MSACAppCenter.h:
```diff
@@ -68,76 +69,45 @@
+ * The flag indicates whether the SDK has already been configured or not.
  */
-+ (BOOL)isConfigured;
+@property(class, atomic, readonly, getter=isConfigured) BOOL configured;
```
The following is arguable but there's actually something tricky in the readonly swift properties. I realized it recently but a "readonly" boolean Swift property **expressed** as an adjective will keep the `is` 😅. At least that's my interpretation of this line: [Uses of Boolean methods and properties should read as assertions about the receiver when the use is nonmutating](https://swift.org/documentation/api-design-guidelines/). Examples: [isContinuous](https://developer.apple.com/documentation/appkit/nscell/1529179-iscontinuous), [isKeyWindow](https://developer.apple.com/documentation/uikit/uiwindow/1621612-iskeywindow), [isCanceled](https://developer.apple.com/documentation/sirikit/inridecompletionstatus/2132104-iscanceled), [isFileURL](https://developer.apple.com/documentation/foundation/nsurl/1408782-isfileurl)...
I could find counter examples basically multiple bool properties of [NSCell](https://developer.apple.com/documentation/appkit/nscell) though 😬.

DmitriyKirakosyan replied:
Agree, the thing is it's not needed to add the `NS_SWIFT_NAME`, afaik uses the `getter` name by default. So this one is going to be `isConfigured` in swift.

DmitriyKirakosyan replied:
Example:
defined as [`@property(class, readonly, nonatomic, getter=isDebuggerAttached) BOOL debuggerAttached;`](https://github.com/microsoft/appcenter-sdk-apple/pull/2201/files#diff-8048fe5774b68eb770d07695708da3b4c6d67cc0a6028c926a1cc7365301ec03R144)

used in swift file as [`AppCenter.isDebuggerAttached`](https://github.com/microsoft/appcenter-sdk-apple/pull/2201/files#diff-95fe3352dd1452ff16b427124eff0d0d891e7b76ae73f2888e2a17c5756b515aR33)

iamclement replied:
Oh interesting, well perfect then :).

iamclement commented on AppCenter/AppCenter/MSACLogger.h:
```diff
@@ -41,4 +42,6 @@
           function:(const char *)function
               line:(uint)line;
 
++ (MSACLogHandler)getCurrentLogHandler;
```
You can remove the "get" in the name because it's obvious that it returns a loghandler.

iamclement replied:
Usually getters don't start with "get" unless they don't return explicitly what they are supposed to return.

DmitriyKirakosyan replied:
Fixed. Also moved the definition to `MSACLoggerInternal.h` and named `logHandler` for the sake of consistency to the setter `setLogHandler`.

iamclement reviewed (changes requested):
Another pass

iamclement commented on AppCenterCrashes/AppCenterCrashes/MSACCrashes.mm:
```diff
@@ -230,6 +229,10 @@ + (BOOL)hasCrashedInLastSession {
 }
 
++ (_Nullable MSACUserConfirmationHandler)userConfirmationHandler {
+  return [[MSACCrashes sharedInstance] userConfirmationHandler];
```
nit: If that's a property we should probably call it as a property.

iamclement commented on AppCenterCrashes/AppCenterCrashes/MSACCrashes.mm:
```diff
@@ -273,6 +276,10 @@ + (void)disableMachExceptionHandler {
 }
 
++ (_Nullable id<MSACCrashesDelegate>)delegate {
+  return [[MSACCrashes sharedInstance] delegate];
```
nit: If that's a property we should probably call it as a property.

iamclement commented on AppCenterCrashes/AppCenterCrashes/WrapperSDKUtilities/MSACWrapperCrashesHelper.h:
```diff
@@ -14,35 +14,35 @@ NS_ASSUME_NONNULL_BEGIN
  * @return The delegate being used by Crashes.
  */
-+ (id<MSACCrashHandlerSetupDelegate>)getCrashHandlerSetupDelegate;
++ (id<MSACCrashHandlerSetupDelegate>)getCrashHandlerSetupDelegate DEPRECATED_MSG_ATTRIBUTE("Use crashHandlerSetupDelegate instead");
```
Any reason to not just remove this API? Isn't it internal? For our wrapper SDKs?

MatkovIvan replied:
This feature was not intended to change Objective-C API, so we decided not to break all wrappers by this.

iamclement commented on AppCenterDistribute/AppCenterDistribute/MSACDistribute.h:
```diff
@@ -35,44 +36,39 @@ typedef NS_ENUM(NSInteger, MSACUpdateTrack) {
    */
   MSACUpdateTrackPrivate = 2
-};
+} NS_SWIFT_NAME(UpdateTrack);
```
We don't need to also swiftify `MSACUpdateAction` ?

iamclement commented on CHANGELOG.md:
```diff
@@ -4,7 +4,8 @@
 
-* **[Breaking change]** In order to avoid a naming conflict with classes in Apple's private frameworks, we now use "MSAC" prefix instead of "MS". Please follow the [migration guide](replace-with-link) for adopting the new naming.
+* **[Feature]** **[Breaking change]** Make SDK API more Swift-friendly. You can find the API updates in the [swift migration table](https://docs.microsoft.com/appcenter/sdk/getting-started/migration/apple-sdk-update#swift-api-changes).
+* **[Breaking change]** In order to avoid a naming conflict with classes in Apple's private frameworks, we now use "MSAC" prefix instead of "MS". Please follow the [migration guide](https://docs.microsoft.com/appcenter/sdk/getting-started/migration/apple-sdk-update) for adopting the new naming.
```
We don't need to go into details maybe just straight to
```suggestion
* **[Breaking change]** The App Center prefix `MS` in Objective-C is consolidated to `MSAC` for collision prevention. Please follow the [migration guide](https://docs.microsoft.com/appcenter/sdk/getting-started/migration/apple-sdk-update) for adopting the new naming.
```

iamclement commented on AppCenterCrashes/AppCenterCrashes/MSACCrashes.h:
```diff
@@ -148,16 +149,14 @@ typedef NS_ENUM(NSUInteger, MSACUserConfirmation) {
  * @see MSACUserConfirmationHandler
  */
-+ (void)setUserConfirmationHandler:(_Nullable MSACUserConfirmationHandler)userConfirmationHandler;
+@property(class, nonatomic) MSACUserConfirmationHandler _Nullable userConfirmationHandler;
```
BTW should we swiftify `MSACUserConfirmationHandler`, it still has the `MSAC` prefix in Swift?

DmitriyKirakosyan replied:
> The public delegates (e.g. `MSACCrashesDelegate`) have more complicated API names I feel like we should do a pass on them. Particularly paying attention to [omit words that merely repeat type information](https://swift.org/documentation/api-design-guidelines/).
> Other than that I have seen all the files I wanted to review :).

Regarding the delegates, I think swift api already looks fine. Eg:
```
- (void)crashes:(MSACCrashes *)crashes didFailSendingErrorReport:(MSACErrorReport *)errorReport withError:(NSError *)error;
```

Resolves to swift as
```
func crashes(_ crashes: Crashes, didFailSending errorReport: ErrorReport, withError error: Error) {
}
```

Here is the place in swift code https://github.com/microsoft/appcenter-sdk-apple/blob/develop/Sasquatch/SasquatchSwift/AppDelegate.swift#L217

Do you think there still should be some additional swiftify changes applied?

iamclement commented on AppCenterCrashes/AppCenterCrashes/MSACCrashes.h:
```diff
@@ -148,16 +149,14 @@ typedef NS_ENUM(NSUInteger, MSACUserConfirmation) {
  * @see MSACCrashesDelegate
  */
-+ (void)setDelegate:(_Nullable id<MSACCrashesDelegate>)delegate;
+@property(class, nonatomic) id<MSACCrashesDelegate> _Nullable delegate;
```
Worth adding `weak` here? At least for explicitly mentioning it and lift customer's doubt and for  consistency with the `MSACDistribute` delegate which is declared weak.

iamclement reviewed (changes requested):
The public delegates (e.g. `MSACCrashesDelegate`) have more complicated API names I feel like we should do a pass on them. Particularly paying attention to [omit words that merely repeat type information](https://swift.org/documentation/api-design-guidelines/).
Other than that I have seen all the files I wanted to review :).

DmitriyKirakosyan commented:
> The public delegates (e.g. `MSACCrashesDelegate`) have more complicated API names I feel like we should do a pass on them. Particularly paying attention to [omit words that merely repeat type information](https://swift.org/documentation/api-design-guidelines/).
> Other than that I have seen all the files I wanted to review :).

Regarding the delegates, I think swift api already looks fine. Eg:
```
- (void)crashes:(MSACCrashes *)crashes didFailSendingErrorReport:(MSACErrorReport *)errorReport withError:(NSError *)error;
```

Resolves to swift as
```
func crashes(_ crashes: Crashes, didFailSending errorReport: ErrorReport, withError error: Error) {
}
```

Here is the place in swift code https://github.com/microsoft/appcenter-sdk-apple/blob/develop/Sasquatch/SasquatchSwift/AppDelegate.swift#L217

Do you think there still should be some additional swiftify changes applied?

MatkovIvan commented on Sasquatch/SasquatchSwift-Extension/ExtensionViewController.swift:
```diff
@@ -21,15 +21,15 @@ class ExtensionViewController: UIViewController, NCWidgetProviding, MSACCrashesD
   }
   
-  func attachments(with crashes: MSACCrashes, for errorReport: MSACErrorReport) -> [MSACErrorAttachmentLog] {
+    func attachments(with crashes: Crashes, for errorReport: ErrorReport) -> [ErrorAttachmentLog] {
```
```suggestion
  func attachments(with crashes: Crashes, for errorReport: ErrorReport) -> [ErrorAttachmentLog] {
```

iamclement commented:
> > The public delegates (e.g. `MSACCrashesDelegate`) have more complicated API names I feel like we should do a pass on them. Particularly paying attention to [omit words that merely repeat type information](https://swift.org/documentation/api-design-guidelines/).
> > Other than that I have seen all the files I wanted to review :).
> 
> Regarding the delegates, I think swift api already looks fine. Eg:
> 
> ```
> - (void)crashes:(MSACCrashes *)crashes didFailSendingErrorReport:(MSACErrorReport *)errorReport withError:(NSError *)error;
> ```
> 
> Resolves to swift as
> 
> ```
> func crashes(_ crashes: Crashes, didFailSending errorReport: ErrorReport, withError error: Error) {
> }
> ```
> 
> Here is the place in swift code https://github.com/microsoft/appcenter-sdk-apple/blob/develop/Sasquatch/SasquatchSwift/AppDelegate.swift#L217
> 
> Do you think there still should be some additional swiftify changes applied?

This one should be fine but have you checked them all? I saw one yesterday that explicitly labeled `ErrorReport`. I'll double check.

iamclement reviewed (changes requested):
One thing, probably the last :). `error` or `withError` somehow is always explicitly labeled (e.g. [NSApplicationdelegate](https://developer.apple.com/documentation/appkit/nsapplicationdelegate), [NSURLSessionDelegate](https://developer.apple.com/documentation/foundation/urlsessiondelegate/1407776-urlsession))

DmitriyKirakosyan commented:
> One thing, probably the last :). `error` or `withError` somehow is always explicitly labeled (e.g. [NSApplicationdelegate](https://developer.apple.com/documentation/appkit/nsapplicationdelegate), [NSURLSessionDelegate](https://developer.apple.com/documentation/foundation/urlsessiondelegate/1407776-urlsession))

Reverted the previous state for these methods, coz swift default conversion did it correctly. The only method that didn't translate automatically was `func crashes(_ crashes: Crashes!, shouldProcessErrorReport errorReport: ErrorReport!) -> Bool`, so I fixed it. Now all delegate method signatures are converted correctly in swift : https://github.com/microsoft/appcenter-sdk-apple/blob/develop/Sasquatch/SasquatchSwift/AppDelegate.swift#L211-L220

Also removed warning in sasquatch apps.

And converted MSACLogHandler and MSACLogMessageProvider to the short versions for swift.

MatkovIvan commented on AppCenterCrashes/AppCenterCrashes/MSACCrashesDelegate.h:
```diff
@@ -63,8 +63,8 @@ NS_SWIFT_NAME(CrashesDelegate)
 
 NS_ASSUME_NONNULL_END
+
+@end
```
Move before `NS_ASSUME_NONNULL_END`

iamclement reviewed (dismissed):
LGTM

MatkovIvan commented on AppCenter/AppCenter/Internals/Context/Device/MSACDeviceTracker.h:
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#import "MSACWrapperSdkInternal.h"
```
Why do we need this new import?

DmitriyKirakosyan replied:
good catch!

DmitriyKirakosyan replied:
```suggestion
```
